numpy/core/src/multiarray/usertypes.c

```diff
@@ -213,6 +213,9 @@ PyArray_RegisterDataType(PyArray_Descr *descr)
         }
     }
 
+    /* Invalidate cached hash value */
+    descr->hash = -1;
```

Comment:
This seems OK to me, the hashing is a hot mess and I am a bit surprised by it not just returning `0` hash for weird dtypes (unpickled, with metadata, double `.newbyteorder()`).

But, it does seem to work out in practice, so as long as the test is fixed and passes I think this looks good.  Thanks!